internal/plugin/metrics.go

```diff
@@ -0,0 +1,303 @@
+/**
+ * Copyright (c) F5, Inc.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"log/slog"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/nginx/agent/v3/internal/bus"
+	"github.com/nginx/agent/v3/internal/config"
+	"github.com/nginx/agent/v3/internal/metrics"
+	"github.com/nginx/agent/v3/internal/metrics/prometheus"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+
+	metricSdk "go.opentelemetry.io/otel/sdk/metric"
+)
+
+const otelServiceName = "nginx-agent"
+
+type (
+	shutdownFunc func()
+
+	DataSource interface {
+		// Starts the data source and begins data collection at a regular interval. Metrics should be pushed to the
+		// MessagePipe using the metrics topic, as they will be handled in the metrics plugin.
+		Start(ctx context.Context, updateInterval time.Duration) error
```

Comment:
Instead of a Start function would a Collect function be better. So instead of each DataSource having its own ticker, the metrics plugin have its own ticker that would call the Collect function for each DataSource. The Collect function would return an array of DataEntry which the metrics plugin could be aggregate together and send to the exporter.

Reply:
Sure, we can do that. That would just mean that we'd have a common refresh interval for all sources instead of a configurable refresh interval per data source but that's likely fine because we are exporting everything at the same rate anyway.

I'll get to work on this!

Reply:
`DataSource` was renamed into a `MetricsProducer`, which can `Produce()` metrics in the way that you describe `Collect` here. I chose `Producer` to make it perhaps more clear that it's following the Producer-Consumer pattern.